### PR TITLE
Summon refactor Mk. 2

### PR DESCRIPTION
**Pull request by Slayer95**
Partial revert of 6acaa82. The handler causes issues because it's actually added by the engine _after_ scenario-specific handlers.

### CODE REVIEW

File: utils/summon.cfg [hunk 1]
```diff
@@ -51,6 +51,12 @@ For more information, choose the 'Summoning Help' rightclick option which appear
 #enddef
 
 #define SUMMON_MENU_ITEM ID DESCRIPTION COST SUMMONER UNIT
+    #arg EXCEPT_FROM
+    C*,K*
```
Comment (ProditorMagnus) [inline utils/summon.cfg:55]: you probably should handle overlays as well, *^C* style

Comment (Slayer95) [reply utils/summon.cfg:55]: Like this: ``C*,K*,*^C*,*^K*`` ?

**Comment by Slayer95** [comment]
PTAL. This commit:

- Removes indeterminacy in execution of ``post summon`` handlers.
- Fixes OOS in MP from usage of ``select`` event.
- Fixes validation of locations.
- Makes locations more flexible (e.g. it adds support for summoning water elementals on water if desired.)
- Allows cancelling summon of {UNIT} through ``try summon {UNIT}`` event.

**Comment by inferno8** [comment]
Very impressive. Is the PR ready for merging in its current state?

**Comment by Slayer95** [comment]
Waiting for confirmation from @ProditorMagnus re: overlays.

File: utils/summon.cfg [hunk 2]
```diff
@@ -91,34 +97,32 @@ For more information, choose the 'Summoning Help' rightclick option which appear
...
-                    [unit]
-                        type={UNIT}
-                        side=$side_number
-                        x,y=$x1,$y1
-                        moves=0
-                        animate=yes
-                    [/unit]
-                    {MODIFY_UNIT x,y=$x1,$y1 attacks_left 0}
-                    {MODIFY_UNIT x,y=$x1,$y1 moves 0}
+                    [sync_variable]
```
Comment (ProditorMagnus) [inline utils/summon.cfg:100]: I am not familiar with sync_variable so just relying on your testing of you changes. Nothing immediately and strongly wrong noticed.

**Comment by Slayer95** [comment]
Alright. Applied some macro tips from another review and rebased.

Ready :]

**Comment by Slayer95** [comment]
By the way, the current location rules re: castles seem a bit too arbitrary to me.
1 Summoner inside tries to summon outside: Forbidden (OK, critical for balance purposes to avoid stalling)
2. Summoner outside tries to summon inside: Allowed (Why?)
3. Summoner inside tries to summon inside: Forbidden (Why?)

IMO, it would be better if summoning be allowed only if both the summoner and summoned would be in the same side of the castle (inside vs outside), i.e. 1 Forbidden, 2 Forbidden, 3 Allowed.

Flavor-sense, it makes sense wholly. Just think of the castle walls as barriers.

Balance-wise, there are several actions/strategies affected.
1. **Defense**: Sieged summoners are allowed to plug holes in their defenses, but are not allowed to attack the invaders that turn. In a standard castle layout, they would remain exposed to attacks (neither the new zone of control shield the summoner.)
2. **Offense**: Summoners are no longed allowed to sneak magical beings into an enemy castle (very cheesy and disrupting strat.). However, they get the chance to summon reinforcements after the castle is broke through, at the cost of their attack; but would that be a wise decision? Player would need to consider whether the opportunity cost is worth it (after breaking through, a low HP summoner might be better served by retreating before summoning reinforcements.)

**Comment by inferno8** [comment]
Interesting. It would change some strategies of Summoners but in a good way, I think. You analysis convinces me. Yeah, we can try these new summoning mechanics. I approve this proposal.

**Comment by Slayer95** [comment]
Great! Then the ``EXCEPT_FROM`` parameter is no longer needed.

**Pull request merged**